### PR TITLE
[Shallow] Printing topography and free surface

### PR DESCRIPTION
**Pull request by miguelmaso**
This PR is an alternative to #4725 and since I don't need to derive GidOutputProcess, it closes #4730 

The old approach was using **one** model part and **one** instance of a derived GidOutputProcess. The derived GidOutputProcess was printing the model part **twice**.

The new approach uses **two** model parts and **two** instances of the standard GidOutputProcess. It produces the same output than #4725 and MultipleFiles flag can provide additional features such as move the mesh (free surface) with the wave or move the topography according to the erosion.

Captura de pantalla de 2019-05-02 16-39-17

For performance reasons, the user can skip the creation of the auxiliary model part and disable the second GidOutputProcess.

### CODE REVIEW

**Comment by miguelmaso** [comment]
it closes #4725

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+{
+    // The new absolute unique_id
+    IndexType unique_id = 0;
+    mOriginNodesIds.clear();
+
+    // Loop the model parts
+    for (const auto& name : mModelPartNames)
+    {
+        ModelPart& model_part = mrModel.GetModelPart(name);
+        for (int i = 0; i < static_cast<int>(model_part.NumberOfNodes()); ++i)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp]: If the loops are not OpenMP parallel (and they can't be in this specific case), the counter type can be `unsigned int` so you don't need the `static_cast<int>()` anymore.

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                auto it_cond = model_part.ConditionsBegin() + i;
+                it_cond->SetId(mOriginConditionsIds[it_cond->Id()]);
+            }
+        }
+    }
+    // Avoiding the possibility to restore the ids twice
+    mOriginConditionsIds.clear();
+}
+
+StringVectorType IdRenumberingProcess::GetRootModelPartNames() const
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp:155]: I have some concerns regarding this method: 
1) Why you can't use the `GetRootModelPart()` method with each one of the model parts of the model?
2) You are iterating all the model parts in the model container. What happens if the model contains other model parts that are not related with the "shallow waters" problem?
3) If you need this method because you can't use the `GetRootModelPart()`, does it work with recursive model parts (e.g. ModelPartA.SubModelPartB.SubSubModelPartC)?

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.h [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    ///@{
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    Model& mrModel;
+    StringVectorType mModelPartNames;
+    std::unordered_map<IndexType, IndexType> mOriginNodesIds;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_processes/id_renumbering_process.h]: I'd rename these member variables to something like `mOriginNodesIdsMap`(until I arrived to this point I didn't realize it is a map).

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 4]
```diff
@@ -63,18 +61,36 @@ namespace Python
         .def("CheckDryPrimitiveVariables", &ShallowWaterVariablesUtility::CheckDryPrimitiveVariables)
         .def("SetDryWetState", &ShallowWaterVariablesUtility::SetDryWetState)
         .def("DeactivateDryElements", &ShallowWaterVariablesUtility::DeactivateDryElements)
-        .def("DefineDryProperties", &ShallowWaterVariablesUtility::DefineDryProperties)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp:66]: I think that this belongs to another PR (just reporting it in case you didn't note it was added to this one).

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 5]
```diff
@@ -0,0 +1,311 @@
+//    |  /           |
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h:1]: Why there is no .cpp?

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+            max_id = std::max(max_id, prop_id);
+            prop_ids.push_back(prop_id);
+        }
+
+        // Creating the auxiliary ids for the dry domain
+        for (auto id : prop_ids)
+        {
+            // Get pointers to the properties and create the dry property
+            Properties::Pointer wet_prop = mrModelPart.pGetProperties(id);
+            Properties::Pointer dry_prop(new Properties(*wet_prop));
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]:
```suggestion
            auto dry_prop = Kratos::make_shared<Properties>(*wet_prop);
```
The shared pointers should now be constructed using the `Kratos::make_shared<Type>()` method.

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+            dry_prop->SetId(++max_id);
+
+            // Add the new properties to the model part and add them to the maps
+            mrModelPart.AddProperties(dry_prop);
+            mWetToDryPropertiesMap[wet_prop->Id()] = max_id;
+            mDryToWetPropertiesMap[dry_prop->Id()] = id;
+        }
+    }
+
+    void AssignDryWetProperties(Flags Flag)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: Would you mind to add the Doxygen documentation? It will help a lot to understand what these methods are doing (basically what are these dry and wet).

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+
+protected:
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: You can remove the protected section if you don't need it to make the file shorter.

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    ///@name Private Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Un accessible methods
+    ///@{
+
+    /// Assignment operator.
+    PostProcessUtilities& operator=(PostProcessUtilities const& rOther);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]:
```suggestion
    PostProcessUtilities& operator=(PostProcessUtilities const& rOther) = delete;
```

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+    /// Copy constructor.
+    PostProcessUtilities(PostProcessUtilities const& rOther);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]:
```suggestion
    PostProcessUtilities(PostProcessUtilities const& rOther) = delete;
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+namespace Kratos
+{
+
+ReplicateModelPartUtility::ReplicateModelPartUtility(
+    ModelPart& rOriginModelPart,
+    ModelPart& rDestinationModelPart,
+    bool ReplicateSubModelParts
+) : mrOriginModelPart(rOriginModelPart)
+  , mrDestinationModelPart(rDestinationModelPart)
+{
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp:35]:
```suggestion
, mReplicateSubModelParts(ReplicateSubModelParts)
{
```
To make the member variable `const bool`

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+            dest_nodes_of_collection[tag].push_back(unique_node_id);
+        }
+    }
+
+    // We also need to replicate the properties before creating the new elements
+    std::unordered_map<IndexType, Properties::Pointer> replicated_properties;
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfProperties(); ++i)
+    {
+        auto it_prop = mrOriginModelPart.PropertiesBegin() + i;
+        Properties::Pointer new_prop(new Properties(*it_prop));
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]:
```suggestion
        auto new_prop = Kratos::make_shared<Properties>(*it_prop);
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        new_prop->SetId(++unique_prop_id);
+        mrDestinationModelPart.AddProperties(new_prop);
+        replicated_properties[it_prop->Id()] = new_prop;
+    }
+
+    // Note: we need to manually create the geometry with the cloned nodes
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfElements(); ++i)
+    {
+        auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+        PointerVector<Node<3>> new_elem_nodes;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]:
```suggestion
        Geometry::PointsArrayType new_elem_nodes;
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+            auto tag = origin_elements_tags[unique_elem_id];
+            dest_elems_of_collection[tag].push_back(unique_elem_id);
+        }
+    }
+
+    // Note: we need to manually create the geometry with the cloned nodes
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfConditions(); i++)
+    {
+        auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+        PointerVector<Node<3>> new_cond_nodes;
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]:
```suggestion
        Geometry::PointsArrayType new_elem_nodes;
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+        PointerVector<Node<3>> new_elem_nodes;
+        for (auto& node : it_elem->GetGeometry())
+        {
+            new_elem_nodes.push_back(mReplicatedNodes[node.Id()]);
+        }
+        auto new_elem = it_elem->Clone(++unique_elem_id, new_elem_nodes);
+        new_elem->SetProperties(replicated_properties[it_elem->GetProperties().Id()]);
+        mrDestinationModelPart.AddElement(new_elem);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: It is faster to save the elements in an auxiliary array to add them altogether using the `AddElements`. Something like:
```
mrDestinationModelPart.AddElements(
    aux_array_with_element_pointers.begin(),
    aux_array_with_element_pointers.end());
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+        PointerVector<Node<3>> new_cond_nodes;
+        for (auto& node : it_cond->GetGeometry())
+        {
+            new_cond_nodes.push_back(mReplicatedNodes[node.Id()]);
+        }
+        auto new_cond = it_cond->Clone(++unique_cond_id, new_cond_nodes);
+        new_cond->SetProperties(replicated_properties[it_cond->GetProperties().Id()]);
+        mrDestinationModelPart.AddCondition(new_cond);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: The same `AddConditions` method is also available.

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+                    model_part.AddElements(dest_elems_of_collection[tag]);
+                    model_part.AddConditions(dest_conds_of_collection[tag]);
+                }
+            }
+        }
+    } // if (mReplicateSubModelParts)
+}
+
+
+void ReplicateModelPartUtility::SetOriginMeshPosition()
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: The method name is misleading. I'd suggest to rename it to `SetOriginMeshZCoordinate`.

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+    #pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(mrOriginModelPart.NumberOfNodes()); ++i)
+    {
+        auto it_node = mrOriginModelPart.NodesBegin() + i;
+        it_node->Z() = it_node->FastGetSolutionStepValue(rVariable);
+    }
+}
+
+
+void ReplicateModelPartUtility::SetDestinationMeshPosition()
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: These methods have exactly the same implementation as the origin mesh ones. I'm wondering if makes sense to pass a reference to a model part as argument to avoid the duplication.

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void Replicate();
+
+    template<class TVarType>
+    void TransferVariable(TVarType& rVariable)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h:97]: I think that there is something in the `VariableUtils` that does this operation. Have you check?

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+        for (int i = 0; i < static_cast<int>(mrOriginModelPart.NumberOfNodes()); ++i)
+        {
+            const auto it_node = mrOriginModelPart.NodesBegin() + i;
+            const auto dest_node = mReplicatedNodes[it_node->Id()];
+            dest_node->FastGetSolutionStepValue(rVariable) = it_node->FastGetSolutionStepValue(rVariable);
+        }
+    }
+
+    template<class TVarType>
+    void TransferNonHistoricalVariable(TVarType& rVariable)
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h:112]: I think that there is something in the `VariableUtils` that does this operation. Have you check?

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+    ///@name Private Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Un accessible methods
+    ///@{
+
+    /// Assignment operator.
+    ReplicateModelPartUtility& operator=(ReplicateModelPartUtility const& rOther);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h]:
```suggestion
    ReplicateModelPartUtility& operator=(ReplicateModelPartUtility const& rOther) = delete;
```

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+
+    /// Copy constructor.
+    ReplicateModelPartUtility(ReplicateModelPartUtility const& rOther);
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h]:
```suggestion
    ReplicateModelPartUtility(ReplicateModelPartUtility const& rOther) = delete;
```

File: applications/ShallowWaterApplication/python_scripts/visualization_mesh_process.py [hunk 23]
```diff
@@ -1,39 +1,97 @@
...
+                self.topography_utility.TransferVariable(variable)
+            for variable in self.nonhistorical_variables:
+                self.topography_utility.TransferNonHistoricalVariable(variable)
+            self.topography_utility.SetDestinationMeshPosition(self.topography_variable)
+            if self.update_free_surface:
+                self.topography_utility.SetOriginMeshPosition(self.free_surface_variable)
+            else:
+                self.topography_utility.SetOriginMeshPosition()
+
+    def ExecuteInitializeSolutionStep(self):
```
Comment (rubenzorrilla) [inline applications/ShallowWaterApplication/python_scripts/visualization_mesh_process.py]: You can avoid this if you assign the same `ProcessInfo` container to the `topographic_model_part` when creating it.

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
In general terms looks fine to me. It is much more modular than the first approach. However, there are some minor concerns that can be addressed before merging.

**Comment by miguelmaso** [comment]
Thanks, I'm going to apply that changes

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp (hunk 1 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp]: Fixed

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp (hunk 2 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_processes/id_renumbering_process.cpp:155]: 1. Now I'm using `GetRootModelPart()`
2. There are two constructors: one with a vector of strings to specify the "shallow water" related model parts. The other constructor calls this method and assumes that all model parts are related to the "shallow water" problem.
3. Solved with 1

File: applications/ShallowWaterApplication/custom_processes/id_renumbering_process.h (hunk 3 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_processes/id_renumbering_process.h]: Renamed

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 4 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp:66]: Belongs to that PR.
I don't need those methods anymore with the new implementation.

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h (hunk 5 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h:1]: Now there is a .cpp

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h (hunk 6 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: Fixed

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h (hunk 7 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: I have added Doxygen to the 3 headers

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h (hunk 8 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: Done

File: applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h (hunk 9 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/post_process_utilities.h]: done

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp (hunk 11 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp:35]: done

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp (hunk 13 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: Done

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp (hunk 15 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: Done

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp (hunk 17 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: Method renamed

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp (hunk 18 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp]: The duplication is avoided

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h (hunk 19 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h:97]: I can't do that with `VariableUtils` because I'm using a map to link the nodes on the origin model part to the nodes on the destination model part.

File: applications/ShallowWaterApplication/python_scripts/visualization_mesh_process.py (hunk 23 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/python_scripts/visualization_mesh_process.py]: Thanks, very useful

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 4 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp:66]: the new utility manages several properties to separate in different layers the topography, the wet domain and the dry domain. The picture represents the run-up of a wave:

Captura de pantalla de 2019-05-03 12-48-50

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h (hunk 19 above)
Comment (rubenzorrilla) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h:97]: OK. My apologies.

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 4 above)
Comment (rubenzorrilla) [reply applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp:66]: Now it is clear to me. Sorry for the inconveniences.

File: applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h (hunk 19 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.h:97]: Any problem, all your suggestions are very useful. Thanks again

**Comment by rubenzorrilla** [comment]
@miguelmaso I think that you already updated everything. Am I wrong?

**Comment by miguelmaso** [comment]
yes, I updated everything and also, some changes suggested by codacy

**Review by rubenzorrilla — APPROVED** [review]
No more concerns from my side then. Thanks!
